oauthenticator/github.py

```diff
@@ -191,8 +193,23 @@ async def authenticate(self, handler, data=None):
         auth_state['access_token'] = access_token
         # store the whole user model in auth_state.github_user
         auth_state['github_user'] = resp_json
-        # A public email will return in the initial query (assuming default scope).
-        # Private will not.
+        # If a public email is not available, an extra API call has to be made
+        # to a /user/emails using the access token to retrieve emails. The
+        # scopes relevant for this are checked based on this documentation:
+        # https://docs.github.com/en/developers/apps/building-oauth-apps/scopes-for-oauth-apps#normalized-scopes
+        if not auth_state['github_user']['email'] and (
+            'user' in granted_scopes or 'user:email' in granted_scopes
+        ):
+            req = HTTPRequest(
+                self.github_api + "/user/emails",
+                method="GET",
+                headers=_api_headers(access_token),
+                validate_cert=self.validate_server_cert,
+            )
+            resp_json = await self.fetch(req, "fetching user emails")
+            for val in resp_json:
+                if val["primary"]:
+                    auth_state['github_user']['email'] = val['email']
```

Comment:
This logic seems weird to me. Is resp_json an list, and the list elements always contain a key called primary that sometimes is a falsy value, and we set the email based on the last list element having a non-falsy "primary" key value?

Reply:
Okay it makes sense, we can add a `break` statement as well after finding the primary email for readability

```json
[
  {
    "email": "email1@example.com",
    "primary": true,
    "verified": true,
    "visibility": "public"
  },
  {
    "email": "email2@example.com",
    "primary": false,
    "verified": true,
    "visibility": null
  }
]
```

Reply:
```suggestion
            for val in resp_json:
                if val["primary"]:
                    auth_state['github_user']['email'] = val['email']
                    break
```